File: solutions/270.closest-binary-search-tree-value/closest-binary-search-tree-value.js
    /**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} target
 * @return {number}
 */
const closestValue = (root, value) => {
  let diff = Number.MAX_SAFE_INTEGER
  let res

  const iter = (node) => {
    if (!node) return

    let abs = Math.abs(node.val - value)

    if (abs < diff) {
      res = node.val
      diff = abs
    }

    iter(node.left)
    iter(node.right)
  }

  iter(root)

  return res
}

File: solutions/270.closest-binary-search-tree-value/closest-binary-search-tree-value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        cur_dit = float('inf')
        ret = None

        def dfs(node):
            nonlocal cur_dit, ret, target
            if not node:
                return

            if abs(node.val - target) < cur_dit:
                ret = node.val
                cur_dit = abs(node.val - target)

            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ret
